### DISCUSSION TITLE
integrate Company portal update

### DISCUSSION THREAD

**Discussion by LorealEndPoint** [macadmins/SupportCompanion#1] [Ideas]
could be good, and especially with the change that will arrive to manage apps through Intune, to have the capability to manage Application updates pushed by the Company portal instead of MSC.

**Comment by almenscorner** [comment]
Would love to add support for Intune pushed app updates and installs but this also requires Microsoft to expose this information on the device.

With Munki all data is stored locally in PLIST files with information about what apps are installed, needs to be updated etc.
Using this info I can then link to a specific app or part in MSC using for example `munki://detail-SomeApp`.

With Intune, none of this information nor links are exposed either through local files or CLI which makes it hard to add support for.

**Reply by LorealEndPoint** [reply]
hum thinking like that, through an API call we should be able to have the information directly in Intune ( get the version of the pkg..... ) and comparing date from Intune and from the Mac, we could have the info....
the bad point is that it will require lots of API call to get the info.... and because MS is changing lots of APIs (Report V2) i guess it will not be efficient....

**Reply by almenscorner** [reply]
Yeah API is possible but I want to keep this light weight and not make complex calls and store authentication info anywhere, simple info like this should be easily accessible locally on the machine

**Reply by LorealEndPoint** [reply]
hum, if i'm not mistaken we have the information into the IntuneMdmDaemon.log 
there are the policy that are pushed, the Macos PKG (not the LOB) that could be a first part?

**Reply by almenscorner** [reply]
If anything you only have the ID available in the logs, it does not really give you information about app name, version what is pending etc?

**Reply by LorealEndPoint** [reply]
you have some info: 
here is a part of my log: 
024-05-13 10:20:12:594 | IntuneMDM-Daemon | I | 12910 | AppPolicyHandler | Handling app policy. PolicyID: c9e76f09-cb24-4d88-9323-9069de59a891, Primary BundleID: menu.nomad.DEPNotify, IgnoreVersion: true, Count: 1, AppType: PKG, App Policy Intent: RequiredInstall
2024-05-13 10:20:13:140 | IntuneMDM-Daemon | I | 7371 | AppDetection | Detecting app with specific bundle ID. PolicyID: c9e76f09-cb24-4d88-9323-9069de59a891, AppName: TS-DEPNotify, BundleID: menu.nomad.DEPNotify, IgnoreVersion: true
2024-05-13 10:20:13:141 | IntuneMDM-Daemon | I | 7371 | AppDetection | Found path URL for PKG app. PolicyID: c9e76f09-cb24-4d88-9323-9069de59a891, AppName: TS-DEPNotify, BundleID: menu.nomad.DEPNotify, IgnoreVersion: true
2024-05-13 10:20:13:144 | IntuneMDM-Daemon | I | 7371 | AppDetection | App with specific bundle ID is installed on the device. PolicyID: c9e76f09-cb24-4d88-9323-9069de59a891, AppName: TS-DEPNotify, BundleID: menu.nomad.DEPNotify, IgnoreVersion: true
2024-05-13 10:20:13:379 | IntuneMDM-Daemon | I | 10558 | AppInstallManager | All apps already installed PolicyID: c9e76f09-cb24-4d88-9323-9069de59a891, AppName: TS-DEPNotify, AppType: PKG, BundleID: menu.nomad.DEPNotify
2024-05-13 10:20:13:381 | IntuneMDM-Daemon | I | 10558 | AppPolicyHandler | Handling app policy finished. PolicyID: c9e76f09-cb24-4d88-9323-9069de59a891, Primary BundleID: menu.nomad.DEPNotify, IgnoreVersion: true, Count: 1, AppType: PKG, App Policy Intent: RequiredInstall
}, "ResultCreatedTimeUTC": 2024-05-13T08:20:13Z, "PolicyId": c9e76f09-cb24-4d88-9323-9069de59a891, "ApplicationName": TS-DEPNotify, "Intent": 3, "PolicyType": 12, "EnforcementStateMessage": {
}, "ApplicationName": TS-DEPNotify, "Intent": 3, "PolicyType": 12, "PolicyId": c9e76f09-cb24-4d88-9323-9069de59a891, "ComplianceStateMessage": {

SO yes you have the Name, the policy ID.
i'm quite sure we can have the version somewhere...... let me search a bit but could be really nice

**Reply by almenscorner** [reply]
Alright then it might be feasible to add depending on how difficult it would be to extract the specific info that is needed. Just wish MS would store this info elsewhere as it is available in a log..

**Reply by almenscorner** [reply]
what's the message in the log if an update is pending to be installed?

**Reply by LorealEndPoint** [reply]
need to check

**Reply by LorealEndPoint** [reply]
One Other information that could be used.
Intune to have the information if the app is installed or not (or actually the state), is using : 
sudo /usr/libexec/mdmclient QueryInstalledApps.

Here is the info about the same app: 
  {
            BundleSize = 15246978;
            Identifier = "menu.nomad.DEPNotify";
            Installing = 0;
            Name = "DEPNotify.app";
            ShortVersion = "1.1.6";
            Version = "1.1.6";
        },

so Now need to update the pkg to check the result with all the command :)

**Reply by almenscorner** [reply]
Right so I am able to get some info from the log that contains this:
```
PolicyResult: [""PolicyId"": bdf7ad6f-5fa2-4cdd-a0f8-542fbeea9de1, ""ComplianceStateMessage"": {
    Applicability = 0;
    ComplianceState = 1;
    DesiredState = 2;
    InstallContext = 2;
    ProductVersion = ""5.15.10 (21826)"";
    TargetType = 1;
}, ""PolicyType"": 12, ""ApplicationName"": Zoom, ""EnforcementStateMessage"": {
    EnforcementState = 1000;
}, ""ResultCreatedTimeUTC"": 2024-05-13T10:54:52Z, ""Intent"": 3, ""PolicyVersion"": 1]
```

Which then I am able to convert to values:
```
Application Name: Zoom
Product Version: 5.15.10 (21826)
```

But, question is, how to determine based on this info if it's a pending update app or what state it's actually in and how often this gets written to the log and how many times we would not be able to grab this info as it's not in the logs etc. etc.

**Reply by almenscorner** [reply]
maybe @theneiljohnson has some idea here

**Reply by theneiljohnson** [reply]
I've asked a couple of our engineers to see if anyone has the numeric state to meanings documented anywhere...

**Reply by almenscorner** [reply]
Thanks Neil!

**Reply by almenscorner** [reply]
This for example 
```2024-05-13 06:39:18:483 | IntuneMDM-Daemon | I | 54465 | AppInstallManager | App running, update deferred PolicyID: d871dec1-f6eb-4b38-b382-067030411c7f, AppName: VLC Media Player, AppType: DMG, BundleID: org.videolan.vlc```

produces an error in both the logs and Intune using the `PolicyResult`, and the version is then also set to `Unknown` and `ComplianceState` to `4`. This could be used to _kind of_ determine install state but I do not like that it is in fact reported as an error and we then have no clue about the `ProductVersion` as it reports "Unknown" for some reason when the update is pending.

In an instance like this I would like to see it logged as some ComplianceState that is not an error but a "pending" state and what the currently installed version is.

**Reply by theneiljohnson** [reply]
That looks like a detection bug, what does it look like in Intune?

**Reply by almenscorner** [reply]
Looks like this and happens when the update is postponed
![Screenshot 2024-05-13 at 15 51 31](https://github.com/almenscorner/SupportCompanion/assets/78877636/c36797f8-8a96-44ea-b135-1a52a9fbd4c9)

**Reply by almenscorner** [reply]
I just don't get why it would be an Error when it will try later
![Screenshot 2024-05-13 at 15 53 41](https://github.com/almenscorner/SupportCompanion/assets/78877636/f048e862-bf9c-4a0b-a134-a607e667bfde)

**Reply by almenscorner** [reply]
Might be able to use `EnforcementState` which is `5000` when the update "failed" and is pending, I do not know however if this state can mean other errors as well. What I could do is if `5000` means "update pending", I could grab each app with the name "Foo" and check if any of them has that state. That then means an update is pending but I need confirmation on that.

**Reply by LorealEndPoint** [reply]
could be usefull :
Cached app policy result matches the current policy result PolicyID: e009a5b8-8a92-46e9-8d79-905e13c4d8ca, PolicyResult: ["ExitCode": -2016214717, "ComplianceStateMessage": {
    Applicability = 0;
    ComplianceState = 4;
    DesiredState = 2;
    ErrorCode = "-2016214717";
    InstallContext = 2;
    ProductVersion = Unknown;
    TargetType = 2;
}, "PolicyType": 12, "PolicyId": e009a5b8-8a92-46e9-8d79-905e13c4d8ca, "EnforcementStateMessage": {
    EnforcementState = 5000;
    ErrorCode = "-2016214717";
    
    
<img width="1098" alt="Screenshot 2024-05-13 at 17 19 54" src="https://github.com/almenscorner/SupportCompanion/assets/107847603/910d0d29-aa2e-4c61-9d54-bb4aa2bea2d1">

**Reply by almenscorner** [reply]
If what you have in the screen shot yields that state in the logs it looks like the logs doesn't really differentiate state between different errors.

So then the question is how we know if the app is in a pending update state or simply has failed the install. I am only interested in showing apps which has a ProductVersion, i.e. an app which has at some point successfully installed

**Reply by LorealEndPoint** [reply]
Yes, I need to find a pkg that request update.....

**Reply by almenscorner** [reply]
I already tried it and the state is the same when an update is pending. So we need some confirmations from Neil and team what to look for :)

**Reply by almenscorner** [reply]
Just an FYI, I am getting some progress on this feature:

![SCAppIntuneMode](https://github.com/almenscorner/SupportCompanion/assets/78877636/75c5999e-ded3-4c67-a5cd-99c7fae983dd)

**Reply by almenscorner** [reply]
I have pushed a new release with support for Intune apps, please try it out and let me know if it works: https://github.com/almenscorner/SupportCompanion/releases/tag/v1.0.1-beta

**Reply by LorealEndPoint** [reply]
First test, and seems to work like a charm!!!!! 
Congrats let me go deeper in the usage, and I will be able to provide you a better answer.

**Reply by almenscorner** [reply]
Great, thanks for testing promptly :)
